libdns.go

```diff
@@ -404,13 +222,13 @@ func RelativeName(fqdn, zone string) string {
 	return rel
 }
 
-// [AbsoluteName] makes “name” into a fully-qualified domain name (FQDN) by
-// prepending it to “zone” and tidying up the dots. For example, an input of
+// [AbsoluteName] makes name into a fully-qualified domain name (FQDN) by
+// prepending it to zone and tidying up the dots. For example, an input of
 // name “sub” and zone “example.com.” will return “sub.example.com.”.
 //
 // Using “@” as the name is the recommended way to represent the root of the
 // zone; however, unlike the [Record] struct, using the empty string "" for the
-// name *is* permitted here, and will be identically to “@”.
+// name *is* permitted here, and will be treated identically to “@”.
 func AbsoluteName(name, zone string) string {
```

Comment:
While porting `libdns/rfc2136` to use the new interfaces in this PR, one problem that I noticed is that calling `AbsoluteName` on an already-absolute domain (with the trailing dot) still appends the zone to the domain name. I think that this should behave like [`path/filepath.Abs`](https://pkg.go.dev/path/filepath#Abs) and do nothing if the domain name is already absolute.

```suggestion
func AbsoluteName(name, zone string) string {
	if zone == "" {
		return strings.Trim(name, ".")
	}
	if name == "" || name == "@" {
		return zone
	}
	if strings.HasSuffix(name, ".") {
		// Already a FQDN, so just return it
		return name
	}
	return name + "." + zone
}
```

Reply:
So that fails a test case, where the name is `"www."` and the zone is `"example.com."` -- should the output really be `www.`? Should the name be considered absolute only if the zone is empty string (and maybe also ".")? Just trying to figure out what makes the most sense.

Reply:
> where the name is `"www."` and the zone is `"example.com."` -- should the output really be `www.`? 

Yes, the trailing dot means that this is already a FQDN, so `www.` is a valid absolute name. There isn't any domain in the root zone called `www.` right now, but it would ostensibly be a valid name. This isn't strictly theoretical—there are actually a few TLDs with `A` records set. Up until very recently [there used to be quite a few of these](https://en.wikipedia.org/wiki/Top-level_domain#Dotless_domains), but `uz` still works:

- [`https://uz./`](https://uz./)
- `curl --insecure -vvv --doh-url 'https://dns.google/dns-query' https://uz./`

> Should the name be considered absolute only if the zone is empty string (and maybe also ".")? Just trying to figure out what makes the most sense.

The main advantage to doing nothing if there's a trailing dot is that `AbsoluteName` becomes idempotent. So with the following example

```go
package main

import (
	"fmt"

	"github.com/libdns/libdns"
)

func main() {
	zone := "example.com."
	rrs := []libdns.CNAME{
		{
			Name:   "one",
			TTL:    3600,
			Target: "two",
		},
		{
			Name:   "three",
			TTL:    3600,
			Target: "four.example.net.",
		},
	}

	for _, rr := range rrs {
		fmt.Println(libdns.AbsoluteName(rr.Target, zone))
	}
}
```

with this change you'd get

```txt
two.example.com.
four.example.net.
```

but _without_ this change, you'd need the following code to get the same output:

```go
package main

import (
	"fmt"
	"strings"

	"github.com/libdns/libdns"
)

func main() {
	zone := "example.com."
	rrs := []libdns.CNAME{
		{
			Name:   "one",
			TTL:    3600,
			Target: "two",
		},
		{
			Name:   "three",
			TTL:    3600,
			Target: "four.example.net.",
		},
	}

	for _, rr := range rrs {
		target := rr.Target
		if !strings.HasSuffix(target, ".") {
			target = libdns.AbsoluteName(rr.Target, zone)
		}
		fmt.Println(target)
	}
}
```

This change would also make `libdns.AbsoluteName` behave similarly to `path/filepath.Abs`, where you can call it on a path that begins with `/` and it will do nothing:

```go
package main

import (
	"fmt"
	"os"
	"path/filepath"
)

func main() {
	os.Chdir("/tmp/")
	paths := []string{
		"folder/file.txt",
		"/etc/passwd",
	}
	for _, path := range paths {
		path, _ = filepath.Abs(path)
		fmt.Println(path)
		// Call it 4 more times
		for i := 0; i < 4; i++ {
			path, _ = filepath.Abs(path)
		}
		fmt.Println(path)
	}
}
```

```txt
/tmp/folder/file.txt
/tmp/folder/file.txt
/etc/passwd
/etc/passwd
```